File: 2021/day15/solution.py
from collections import defaultdict
import heapq

class Grid:

    def __init__(self, numbers):
        self.grid = [ [0]*len(numbers[0]) for _ in range(len(numbers)) ]      
        for r in range(len(self.grid)):
            for c in range(len(self.grid[0])):
                self.grid[r][c] = numbers[r][c]

        self.costs = defaultdict(int)
        
    
    def optimal_path(self) -> int:
        min_heap, visited = [(0, 0, 0)], set()
        heapq.heapify(min_heap)
        
        while min_heap:
            local_cost, r, c = heapq.heappop(min_heap)
            if (r,c) in visited:
                continue
            
            visited.add((r,c))
            self.costs[(r,c)] = local_cost
            if r == len(self.grid)-1 and c == len(self.grid[0]):
                break
            
            adj = [(1,0), (0,1), (-1,0), (0,-1)]
            for x,y in adj:
                rr, cc = r+x, c+y
                if not (0 <= rr < len(self.grid) and 0 <= cc < len(self.grid[0])):
                    continue
                
                
                heapq.heappush(min_heap, (local_cost+self.grid[rr][cc], rr, cc))

        return self.costs[len(self.grid)-1, len(self.grid[0])-1]
                
                            

            
        
        
                   
def create_grid() -> Grid:
    with open('day15/input.txt') as f:
        lines = f.readlines()
        numbers = []
        for l in lines:
            l = l.strip()
            
            row = []
            for char in l:

                row.append(int(char))
            numbers.append(row)
        return Grid(numbers=numbers)

def part1() -> int:
    grid = create_grid()
    return grid.optimal_path()

def part2() -> int:
    with open("day15/input.txt") as fin:
        raw_data = fin.read().strip()
        map = [[int(i) for i in line] for line in raw_data.split("\n")]
        
        
        N = len(map)
        M = len(map[0])
        
        rows = N * 5
        cols = M * 5
        
        
        def get(r, c):
            x = (map[r % N][c % M] +
                 (r // N) + (c // M))
            return (x - 1) % 9 + 1
        
        
        cost = defaultdict(int)
        
        pq = [(0, 0, 0)]
        heapq.heapify(pq)
        visited = set()
        
        while len(pq) > 0:
            c, row, col = heapq.heappop(pq)
        
            if (row, col) in visited:
                continue
            visited.add((row, col))
        
            cost[(row, col)] = c
        
            if row == rows - 1 and col == cols - 1:
                break
        
            for dr, dc in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
                rr = row + dr
                cc = col + dc
                if not (0 <= rr < rows and 0 <= cc < cols):
                    continue
        
                heapq.heappush(pq, (c + get(rr, cc), rr, cc))
        
        
        print(cost[(rows - 1, cols - 1)])

File: 2021/day15/solution.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <queue>
#include <set>

using namespace std;

vector<int> getIncrementedVector(const vector<int>& vec) {
    vector<int> incrementedVector = {};
    for (const int& v : vec) {
        auto val = v + 1;
        if (val > 9) {
            val = 1;
        }
        incrementedVector.push_back(val);
    }
    return incrementedVector;
}

vector<vector<int>> getExtendedGrid(const vector<vector<int>>& grid, int scaling) {
    vector<vector<int>> extendedGrid = {};

    // extend the grid to contain all the original rows but with extended columns
    for (const auto & r : grid) {
        vector<vector<int>> rows = {};
        rows.push_back(r);
        for (auto s = 0; s < scaling-1; s++) {
            auto chunk = getIncrementedVector(rows[rows.size()-1]);
            rows.push_back(chunk);
        }

        vector<int> extendedRow = {};
        for (const auto& row : rows) {
            for (auto val : row) {
                extendedRow.push_back(val);
            }
        }
        extendedGrid.push_back(extendedRow);
    }

    // extend the columns downward according to the pattern
    vector<vector<vector<int>>> chunks = {};
    chunks.push_back(extendedGrid);
    for (auto s = 0; s < scaling-1; s++) {
        vector<vector<int>> nextChunk = {};
        for (const auto& chunk : chunks[chunks.size()-1]) {
            nextChunk.push_back(getIncrementedVector(chunk));
        }
        chunks.push_back(nextChunk);
    }
    for (int i = 1; i < chunks.size(); i++) {
        for (const auto& row : chunks[i]) {
            extendedGrid.push_back(row);
        }
    }

    return extendedGrid;
}

vector<vector<int>> getGrid(const string& filePath) {
    ifstream inputFile(filePath);
    vector<string> lines;
    if (!inputFile) {
        cerr << "Failed to open input file." << endl;
        return {};
    }
    string line;
    while (getline(inputFile, line)) {
        lines.push_back(line);
    }
    vector<vector<int>> grid = {};
    for (const auto& l : lines) {
        vector<int> row = {};
        for (auto c : l) {
            // Zero must be subtracted to account for ASCII conversion
            int val = c - '0';
            row.push_back(val);
        }
        grid.push_back(row);
    }
    return grid;
}

struct path {
    int cost;
    int r;
    int c;

    bool operator<(const path& other) const {
        return cost > other.cost;
    }
};

int lowestPathCost(vector<vector<int>> grid) {
    priority_queue<path> minHeap;
    minHeap.push(path{0, 0, 0});
    set<pair<int,int>> visited;
    while (!minHeap.empty()) {
        path p = minHeap.top();
        minHeap.pop();
        int totalCost = p.cost;
        int r = p.r;
        int c = p.c;

        if (visited.count(make_pair(r,c)) > 0) {
            continue;
        }
        visited.insert(make_pair(r,c));

        if (r != 0 || c != 0) {
            totalCost += grid[r][c];
        }
        if (r == grid.size()-1 && c == grid[0].size()-1) {
            return totalCost;
        }
        path nextPaths[] = {
                {totalCost, r-1, c},
                {totalCost, r+1, c},
                {totalCost, r, c-1},
                {totalCost, r, c+1},
        };

        for (const path& nextP : nextPaths) {
            auto key = make_pair(nextP.r, nextP.c);
            if (nextP.r >= 0 && nextP.c >= 0 && nextP.r < grid.size() && nextP.c < grid[0].size() && visited.count(key) == 0) {
                minHeap.push(nextP);
            }
        }
    }
    return -1;
}


int main() {
    auto grid = getGrid("input.txt");
    auto extendedGrid = getExtendedGrid(grid, 5);
    
    cout << "Part 1: " + to_string(lowestPathCost(grid)) << endl;
    cout << "Part 2: " + to_string(lowestPathCost(extendedGrid)) << endl;
}